lib/spack/spack/compiler.py

```diff
@@ -286,26 +298,115 @@ def __str__(self):
                 str(self.operating_system)))))
 
 
-def _get_versioned_tuple(compiler_check_tuple):
-    full_path, prefix, suffix, detect_version = compiler_check_tuple
+@functools_backport.total_ordering
+class _CompilerID(collections.namedtuple('_CompilerIDBase', [
+    'os', 'cmp_cls', 'version'
+])):
```

Comment:
> If anything they make this more heavyweight... namedtuples already define total ordering and hash semantics...

So, my first try at this was just:
```python
_CompilerID = collections.namedtuple('_CompilerID', ['os', 'cmp_cls', 'version'])
```
but this clashes with the fact that we are passing around a *class* for compilers, not an instance of a class. This means we can't order `_CompilerID` because:
```python
TypeError: '<' not supported between instances of 'type' and 'type'
```
The other part is that the compiler class is unhashable, as it contains a lot of unhashable types. 

What I added here is just a workaround to be able to order on `_CompilerID` + be able to insert a `_CompilerID` into a hash based container. This is basically because uniqueness and ordering seemed important in the original algorithm used to detect compilers.

Reply:
Giving a better look right now at the stuff in `spack.compilers` module. I see that in fact we have all the functions we need there. I can remove all the complexity and just pass the compiler name. Then I can use:
```python
compiler_cls = spack.compilers.class_for_compiler_name(name)
```
in client code.

Reply:
Done in 5448ef0